File: frontend-app/src/api/simulacion.js
/**
 * Cliente API para el backend de simulación.
 * Todas las llamadas pasan por el proxy de Vite (/api → http://localhost:8000/api).
 */

const BASE = '/api'

/**
 * Verifica que el backend esté activo.
 * @returns {{ status: string, mensaje: string }}
 */
export async function healthCheck() {
  const res = await fetch(`${BASE}/health`)
  if (!res.ok) throw new Error('Backend no disponible')
  return res.json()
}

/**
 * Ejecuta la simulación enviando los parámetros al backend.
 *
 * @param {{
 *   replicaciones: number,
 *   tiempo_simulacion: number,
 *   tasa_llegada: number,
 *   semilla: number
 * }} params
 *
 * @returns {Promise<{
 *   throughput: number,
 *   tiempo_total: number,
 *   grupos: Array<{id, maquinas, utilizacion, espera_promedio, cola_promedio}>,
 *   comparativa: Array<{grupo, throughput, mejora_pct, espera_g5}>
 * }>}
 */
export async function simular(params) {
  const res = await fetch(`${BASE}/simular`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })

  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || `Error ${res.status}`)
  }

  return res.json()
}

/**
 * Genera números pseudoaleatorios con el LCG.
 * @param {{ semilla, cantidad, modulo, multiplicador, incremento }} params
 * @returns {{ numeros_crudos, numeros_normalizados, cantidad }}
 */
export async function generar(params) {
  const res = await fetch(`${BASE}/generar`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || `Error ${res.status}`)
  }
  return res.json()
}

/**
 * Verifica que los parámetros del LCG sean válidos (Hull-Dobell).
 * @param {{ semilla, cantidad, modulo, multiplicador, incremento }} params
 * @returns {{ valido, modulo, multiplicador, incremento }}
 */
export async function verificarParametros(params) {
  const res = await fetch(`${BASE}/verificar-parametros`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || `Error ${res.status}`)
  }
  return res.json()
}

/**
 * Dado solo la cantidad, devuelve parámetros LCG válidos generados automáticamente.
 * @param {number} cantidad
 * @returns {{ semilla, cantidad, modulo, multiplicador, incremento, _info }}
 */
export async function generarParametros(cantidad) {
  const res = await fetch(`${BASE}/generar-parametros/${cantidad}`)
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || `Error ${res.status}`)
  }
  return res.json()
}

File: backend/routes/simulacion.py
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Any
from services import simulador
from services import generadorNumeros
import random

router = APIRouter()

class ParametrosSimulacion(BaseModel):
    replicaciones: int = 10
    tiempo_simulacion: float = 480  
    semilla: int = 42
    tasa_llegada: float = 20


class ResultadoGrupo(BaseModel):
    id: int
    maquinas: int
    utilizacion: float        
    espera_promedio: float   
    cola_promedio: float


class ResultadoSimulacion(BaseModel):
    throughput: float
    tiempo_total: float
    grupos: list[ResultadoGrupo]
    comparativa: list[Any]
    log_eventos: list[Any] = []  


class ParametrosGenerador(BaseModel):
    semilla: int
    cantidad: int
    modulo: int
    multiplicador: int
    incremento: int




@router.post("/simular", response_model=ResultadoSimulacion)
def simular(params: ParametrosSimulacion):
    resultado = simulador.ejecutar(params)
    return resultado



@router.post("/generar")
def generar(params: ParametrosGenerador):
    numeros = generadorNumeros.generar_numeros_congurencia_lineal(
        params.semilla,
        params.cantidad,
        params.modulo,
        params.multiplicador,
        params.incremento,
    )
    normalizados = [n / params.modulo for n in numeros]
    return {
        "numeros_crudos": numeros,
        "numeros_normalizados": normalizados,
        "cantidad": len(numeros),
    }


@router.post("/verificar-parametros")
def verificar_parametros(params: ParametrosGenerador):

    resultado = generadorNumeros.verificador_parametros_generador(
        params.semilla,
        params.cantidad,
        params.modulo,
        params.multiplicador,
        params.incremento,
    )
    return resultado


@router.get("/generar-parametros/{cantidad}")
def generar_parametros(cantidad: int):
    return generadorNumeros.generar_parametros_apartir_de_cantidad(cantidad)

@router.get("/semilla-aleatoria")
def semilla_aleatoria():
    return {"semilla": random.randint(1, 99999)}
